### PR TITLE
Machine removal

### PR DESCRIPTION
**Pull request by huwshimi**
Allow machines and containers to be removed.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- use the flag /:flags:/mv/
- add a service
- open the machine view
- auto place the unit
- open the context menu on the machine and click destroy
- the machine should be destroyed

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3786/
Test FAILed.

File: jujugui/static/gui/src/app/models/models.js [hunk 1]
```diff
@@ -2757,13 +2757,7 @@ YUI.add('juju-models', function(Y) {
...
       var changedMachines = [];
       this.machines.each(function(machine) {
-        var units = machine.units;
-        if (!units) {
-          // If the machine view hasn't been opened yet then there won't be a
-          // units collection on the machines. So we need to get the 'real'
-          // unit list.
-          units = this.units.filterByMachine(machine.id, true);
-        }
+        var units = this.units.filterByMachine(machine.id, true);
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/models/models.js]: Removed this kinda hacky unit list, we can just do a lookup using the machine id.

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 2]
```diff
@@ -711,7 +711,8 @@ YUI.add('environment-change-set', function(Y) {
       var db = this.get('db');
       var machine = db.machines.getById(command.args[0]);
       var removedUnits = [];
-      machine.units.forEach(function(unit) {
+      var units = db.units.filterByMachine(machine.id);
+      units.forEach(function(unit) {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: As above, removed this kinda hacky unit list which relied on the machine view having been opened before a machine was destroyed, we can just do a lookup using the machine id.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3787/
Test PASSed.

**Comment by hatched** [comment]
QA Issue:
 
- I auto placed a unit and then before committing destroyed the machine. It left the UI representation for the 'root container' in the right column. Closing and re-opening the mv fixed it.

- There is no UI update to the machine when it is pending destroy. I assumed this would turn blue or something.

- If I commit the machines and then destroy it. It leaves both the machine and root container visible but with no units in it. Closing and re-opening the mv does not fix this. It appears that it does not get deleted.

**Comment by huwshimi** [comment]
@hatched it looks like we were never listening to machine events correctly. I've fixed that first QA issue. There is a card to add uncommitted states in the MV. The last issue is related to the email thread about properly destroying machines from the GUI.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3810/
Test PASSed.

File: jujugui/static/gui/src/app/widgets/autodeploy-extension.js [hunk 3]
```diff
@@ -75,7 +75,6 @@ YUI.add('autodeploy-extension', function(Y) {
         parentId: parentId,
         constraints: constraints || {}
       }], callback, {modelId: machine.id});
-      db.fire('update');
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/widgets/autodeploy-extension.js]: This is a change I had previously made to make the UI update on machine creation, this has now been replaced with the listener in app.js.

**Comment by hatched** [comment]
QA :ok:
Thanks for making that fix!

**Comment by hatched** [comment]
:+1:

**Comment by huwshimi** [comment]
Thanks for the review @hatched :shipit:!

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**